windows/ReactTestApp/MainPage.cpp

```diff
@@ -6,27 +6,92 @@
+        std::optional<::ReactTestApp::Manifest> manifest = ::ReactTestApp::GetManifest();
+        if (manifest == std::nullopt) {
+            MenuFlyoutItem newMenuItem;
+            newMenuItem.Text(L"Couldn't parse app.json");
+            newMenuItem.IsEnabled(false);
+            menuItems.Append(newMenuItem);
+        } else {
+            for (auto &&c : manifest.value().components) {
+                hstring componentDisplayName = to_hstring(c.displayName.value_or(c.appKey));
+                hstring componentName = to_hstring(c.appKey);
+                ReactTestApp::ComponentViewModel newComponent =
+                    winrt::make<ComponentViewModel>(componentName, componentDisplayName);
+                m_components.push_back(newComponent);
+
+                MenuFlyoutItem newMenuItem;
+                newMenuItem.CommandParameter(newComponent);
+                newMenuItem.Text(newComponent.DisplayName());
+                newMenuItem.Click({this, &MainPage::SetReactComponentName});
+                menuItems.Append(newMenuItem);
+            }
+        }
+    }
+
+    void MainPage::InitReact()
+    {
+        // TODO fallback to JS bundle
+        LoadJSBundleFrom(JSBundleSource::DevServer);
+
+        // If only one component is present load it automatically
+        if (m_components.size() == 1) {
+            ReactRootView().ComponentName(m_components.at(0).AppKey());
         }
 
-        return components;
+        ReactRootView().ReactNativeHost(m_reactNativeHost);
+    }
+
+    void MainPage::LoadJSBundleFrom(JSBundleSource source)
+    {
+        m_reactNativeHost.InstanceSettings().UseLiveReload(source == JSBundleSource::DevServer);
+        m_reactNativeHost.InstanceSettings().UseWebDebugger(source == JSBundleSource::DevServer);
+        m_reactNativeHost.InstanceSettings().UseFastRefresh(source == JSBundleSource::DevServer);
+
+        switch (source) {
+            case JSBundleSource::DevServer:
+                m_reactNativeHost.InstanceSettings().JavaScriptMainModuleName(
+                    L"index.windows");
+                break;
+            case JSBundleSource::Embedded:
+                m_reactNativeHost.InstanceSettings().JavaScriptBundleFile(L"main.windows");  // TODO
+                break;
+        }
+
+        m_reactNativeHost.ReloadInstance();
+    }
+
+    void MainPage::LoadFromJSBundle(Windows::Foundation::IInspectable const &,
+                                    Windows::UI::Xaml::RoutedEventArgs)
+    {
+        LoadJSBundleFrom(JSBundleSource::Embedded);
+    }
+
+    void MainPage::LoadFromDevServer(Windows::Foundation::IInspectable const &,
+                                     Windows::UI::Xaml::RoutedEventArgs)
+    {
+        LoadJSBundleFrom(JSBundleSource::DevServer);
+    }
+
+    void MainPage::SetReactComponentName(Windows::Foundation::IInspectable const &sender,
+                                         Windows::UI::Xaml::RoutedEventArgs e)
```

Comment:
Nit: If `e` is unused, we should not name it.
```suggestion
                                         Windows::UI::Xaml::RoutedEventArgs)
```